### PR TITLE
refactor[ReactDebugHooks]: support parsing stack frames when sourcemaps are available

### PR DESCRIPTION
**Pull request by hoxyq**
Unblocks https://github.com/facebook/react/pull/26446.

## TL;DR:
- The main difference is `isReactWrapper` function implementation. Previously, we would just look at hook stack trace entry and check if `functionName` ends with the corresponding hook primitive name. This doesn't work with the same setup, but with sourcemaps also being present. We use Proxy for handling custom errors, here is the implementation of the Proxy for Dispatcher
https://github.com/facebook/react/blob/666a6bc08589350f77e27d84b9ec26cce5500dd7/packages/react-debug-tools/src/ReactDebugHooks.js#L354-L373

For each supported hook, we have a stub, which saves the stack trace via `new Error()` call. Here is `useState`, for example:
https://github.com/facebook/react/blob/666a6bc08589350f77e27d84b9ec26cce5500dd7/packages/react-debug-tools/src/ReactDebugHooks.js#L142

At runtime, current dispatcher can be this Proxy object - `DispatcherProxy`, this will modify the stack trace. Here is how the corresponding stack frame looks like without sourcemaps:
```javascript
{
        columnNumber: 171,
        lineNumber: 16,
        fileName: '.../build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.min.js',
        functionName: 'Proxy.useState',
        source: '    at Proxy.useState (.../react/build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.min.js:16:171)'
}
```

When sourcemaps are present, this particular frame is different, it no longer mentions primitive hook name.
```javascript
{
        columnNumber: 21,
        lineNumber: 114,
        fileName: '.../build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.js',
        functionName: 'Proxy.Error',
        source: '    at Proxy.Error (.../build/oss-stable/react-debug-tools/cjs/react-debug-tools.production.js:114:21)'
}
```

We do this parsing to skip React's frames, in this implementation we are only looking for user-defined hooks. In order to make this logic sourcemaps independent, instead of looking at stack trace of the error, which is created at runtime (basically once hook is called), we are also looking at primitive hook's stacktraces, which are defined here:
https://github.com/facebook/react/blob/666a6bc08589350f77e27d84b9ec26cce5500dd7/packages/react-debug-tools/src/ReactDebugHooks.js#L57-L99

Notice that we are using general Dispatcher here and not DispatcherProxy. Both with or without sourcemaps stack traces contain corresponding primitive hook name in the `functionName`.

## Also in these changes:
- Added Flow typings for `error-stack-parser` via `flow-typed`, updated other parts of the code which were using this library accordingly.

### CODE REVIEW

**Comment by motiz88** [comment]
@hoxyq and I discussed this offline. I think we should unblock https://github.com/facebook/react/pull/26446 differently.

The source mapping behaviour that breaks the existing logic only happens in the test environment, which suggests that https://github.com/facebook/react/pull/26446 creates divergence between the test and deployed environments of React DevTools. Ergo, AFAICT, we don't have the right kind of test coverage to validate this change, and should instead be looking at ways to keep the tests from breaking post-https://github.com/facebook/react/pull/26446 *without changing any DevTools logic* (e.g. by disabling source map generation for `react-debug-tools`).

**Comment by markerikson** [comment]
@motiz88 anything you need from me to update #26446 ?  I _think_ I'm reading this more as "we should tweak the existing tests" or something along those lines, which suggests I don't need to alter the PR, but wanted to check.

**Comment by hoxyq** [comment]
> @motiz88 anything you need from me to update #26446 ? I _think_ I'm reading this more as "we should tweak the existing tests" or something along those lines, which suggests I don't need to alter the PR, but wanted to check.

We should continue to manually exclude `react-debug-tools` from having sourcemaps. The main point here is that even if we had them, they won't be included in React DevTools build with this artifact, so current tests will diverge from the real world use case.

For now lets keep merge your sourcemaps PR with manually excluding sourcemaps, which I will remove later when move `react-debug-tools` package to React DevTools' backend.
